src/compiler/scala/tools/nsc/typechecker/RefChecks.scala

```diff
@@ -390,90 +402,93 @@ abstract class RefChecks extends Transform {
-                }
+              def exempted = javaDetermined(member) || member.overrides.exists(javaDetermined)
+              // warn that nilary member matched nullary other, so either it was adapted by namer or will be silently mixed in by mixin
+              def warnAdaptedNullaryOverride(): Unit = {
+                val named = if (isMemberClass) "" else s" (${member.fullLocationString})"
+                val msg = s"method$named without a parameter list overrides a method with a single empty one"
+                val namePos = member.pos
+                val action =
+                  if (namePos.isDefined && currentUnit.sourceAt(namePos) == member.decodedName)
+                    runReporting.codeAction("add empty parameter list", namePos.focusEnd, "()", msg)
+                  else Nil
+                overrideErrorOrNullaryWarning(msg, action)
               }
-              else if (other.paramss.isEmpty && !member.paramss.isEmpty &&
-                !javaDetermined(member) && !member.overrides.exists(javaDetermined) &&
-                !member.hasAnnotation(BeanPropertyAttr) && !member.hasAnnotation(BooleanBeanPropertyAttr) &&
-                member.pos.isDefined  // scala/bug#12851
-              ) {
-                val msg = "method with a single empty parameter list overrides method without any parameter list"
-                val namePos = member.pos.focus.withEnd(member.pos.point + member.decodedName.length)
+              def warnExtraParens(): Unit = {
+                val named = if (isMemberClass) "" else s" (${member.fullLocationString})"
+                val msg = s"method$named with a single empty parameter list overrides method without any parameter list"
+                val namePos = member.pos
                 val action =
-                  if (currentUnit.sourceAt(namePos) == member.decodedName)
+                  if (namePos.isDefined && currentUnit.sourceAt(namePos) == member.decodedName)
                     runReporting.codeAction("remove empty parameter list", namePos.focusEnd.withEnd(namePos.end + 2), "", msg, expected = Some(("()", currentUnit)))
                   else Nil
-                if (currentRun.isScala3)
-                  overrideErrorWithMemberInfo(msg, action)
-                else
-                  refchecksWarning(member.pos, msg, WarningCategory.OtherNullaryOverride, action)
+                overrideErrorOrNullaryWarning(msg, action)
+              }
+              if (member.hasAttachment[NullaryOverrideAdapted.type]) {
+                if (!exempted)
+                  warnAdaptedNullaryOverride()
+              }
+              else if (member.paramLists.isEmpty) {
+                // NullaryOverrideAdapted is only added to symbols being compiled, so check for a mismatch
+                // if both symbols are mixed in from the classpath
+                if (!member.isStable && other.paramLists.nonEmpty && !exempted && !other.isJavaDefined)
```

Comment:
Instead of `!other.isJavaDefined`, add `javaDetermined(other)` to `def exempted`?